File: python/src/driver.py
from multiprocessing import Pool, cpu_count
from matrix_generator import generate_matrix
from block_utils import split_into_blocks
from mapper import mapper
from reducer import reducer
from utils import shuffle, write_csv_row
import time

def distributed_block_mm(n, m, p, block_size, workers=None):
    timings = {}

    start = time.time()
    A = generate_matrix(n, m)
    B = generate_matrix(m, p)
    timings["generation"] = time.time() - start

    start = time.time()
    A_blocks = split_into_blocks(A, block_size)
    B_blocks = split_into_blocks(B, block_size)
    timings["blocking"] = time.time() - start

    map_inputs = []
    n_blocks = n // block_size
    m_blocks = m // block_size
    p_blocks = p // block_size

    for (i, k), block in A_blocks.items():
        map_inputs.append((block, "A", i, k, n_blocks, m_blocks, p_blocks))

    for (k, j), block in B_blocks.items():
        map_inputs.append((block, "B", k, j, n_blocks, m_blocks, p_blocks))

    start = time.time()
    with Pool(workers) as pool:
        mapped = pool.starmap(mapper, map_inputs)
    timings["map"] = time.time() - start

    start = time.time()
    mapped_flat = [item for sublist in mapped for item in sublist]
    grouped = shuffle(mapped_flat)
    timings["shuffle"] = time.time() - start

    start = time.time()
    results = []
    for key, values in grouped.items():
        results.append(reducer(key, values, block_size))
    timings["reduce"] = time.time() - start

    timings["total"] = sum(timings.values())

    return results, timings

def run_experiment(n, block_size, workers):
    results, timings = distributed_block_mm(
        n, n, n, block_size, workers
    )

    csv_header = [
        "language", "matrix_size", "block_size", "workers",
        "generation", "blocking", "map", "map_shuffle", "reduce", "total"
    ]

    csv_row = [
        "python", n, block_size, workers,
        f"{timings['generation']:.6f}",
        f"{timings['blocking']:.6f}",
        f"{timings['map']:.6f}",
        f"{timings['shuffle']:.6f}",
        f"{timings['reduce']:.6f}",
        f"{timings['total']:.6f}"
    ]

    write_csv_row("benchmarks/results.csv", csv_header, csv_row)
    print(f"Experiment completed. Results written to benchmarks/results.csv")



if __name__ == "__main__":
    for n in [256, 512, 1024]:
        for block_size in [16, 32, 64]:
            run_experiment(
                n=n,
                block_size=block_size,
                workers=cpu_count()
            )

File: java/src/Driver.java
import java.util.*;
import java.util.concurrent.*;

public class Driver {

    public static void main(String[] args) throws Exception {

        int workers = Runtime.getRuntime().availableProcessors();

        int[] sizes = new int[] {256, 512, 1024};
        int[] blockSizes = new int[] {16, 32, 64};

        String csvPath = "benchmarks/results.csv";
        String header = String.join(",",
                "language", "matrix_size", "block_size", "workers",
        "generation", "blocking", "map", "map_shuffle", "reduce", "total"
        );

        System.out.println("=== Java Distributed Matrix Multiplication Benchmarks ===");

        for (int n : sizes) {
            for (int blockSize : blockSizes) {
                Map<String, Double> timings = new LinkedHashMap<>();

                long t0 = Utils.now();
                double[][] A = MatrixGenerator.generate(n, n);
                double[][] B = MatrixGenerator.generate(n, n);
                long t1 = Utils.now();
                timings.put("generation", Utils.elapsedSeconds(t0, t1));

                Map<BlockIndex, double[][]> Ablocks = BlockUtils.split(A, blockSize);
                Map<BlockIndex, double[][]> Bblocks = BlockUtils.split(B, blockSize);
                long t2 = Utils.now();
                timings.put("blocking", Utils.elapsedSeconds(t1, t2));

                int nBlocks = n / blockSize;

                ExecutorService executor = Executors.newFixedThreadPool(workers);
                List<MapOutput> mappedOutputs = Collections.synchronizedList(new ArrayList<>());

                long tMapStart = Utils.now();

                for (Map.Entry<BlockIndex, double[][]> entry : Ablocks.entrySet()) {
                    final BlockIndex key = entry.getKey();
                    final double[][] value = entry.getValue();
                    executor.submit(() -> {
                        List<MapOutput> outputs = Mapper.map(
                                value,
                                "A",
                                key.i,
                                key.j,
                                nBlocks,
                                nBlocks
                        );
                        mappedOutputs.addAll(outputs);
                    });
                }

                for (Map.Entry<BlockIndex, double[][]> entry : Bblocks.entrySet()) {
                    final BlockIndex key = entry.getKey();
                    final double[][] value = entry.getValue();
                    executor.submit(() -> {
                        List<MapOutput> outputs = Mapper.map(
                                value,
                                "B",
                                key.i,
                                key.j,
                                nBlocks,
                                nBlocks
                        );
                        mappedOutputs.addAll(outputs);
                    });
                }

                executor.shutdown();
                executor.awaitTermination(1, TimeUnit.HOURS);
                long tMapEnd = Utils.now();
                timings.put("map", Utils.elapsedSeconds(tMapStart, tMapEnd));

                long tShuffleStart = Utils.now();
                Map<BlockIndex, List<BlockValue>> shuffled = new HashMap<>();
                for (MapOutput o : mappedOutputs) {
                    shuffled
                        .computeIfAbsent(o.key, k -> new ArrayList<>())
                        .add(o.value);
                }
                long tShuffleEnd = Utils.now();
                timings.put("shuffle", Utils.elapsedSeconds(tShuffleStart, tShuffleEnd));

                Map<BlockIndex, double[][]> result = new HashMap<>();

                for (Map.Entry<BlockIndex, List<BlockValue>> entry : shuffled.entrySet()) {
                    result.put(
                            entry.getKey(),
                            Reducer.reduce(entry.getValue(), blockSize)
                    );
                }

                long t4 = Utils.now();
                timings.put("reduce", Utils.elapsedSeconds(tShuffleEnd, t4));
                timings.put("total", Utils.elapsedSeconds(t1, t4));

                String row = String.format(java.util.Locale.US,
                    "java,%d,%d,%d,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f",
                        n,
                        blockSize,
                        workers,
                        timings.get("generation"),
                        timings.get("blocking"),
                        timings.get("map"),
                        timings.get("shuffle"),
                        timings.get("reduce"),
                        timings.get("total")
                );

                CSVUtils.writeRow(csvPath, header, row);
                System.out.println("Matrix size: " + n + "x" + n + ", block " + blockSize + ", workers " + workers);
                timings.forEach((k, v) -> System.out.printf(java.util.Locale.US, "%-15s : %.4f s%n", k, v));
                System.out.println("Benchmark results written to " + csvPath);
                System.out.println();
            }
        }
        System.out.println("Benchmarking complete.");
    }
}
